hy/scoping.py

```diff
@@ -0,0 +1,243 @@
+            elif isinstance(self.node, ast.MatchMapping):
+                self.node.rest = new_name
+        self.node.id = new_name
+
+
+class Scope:
+    def __init__(self, stack, scope_id):
+        self.stack = stack
+        self.id = scope_id
+
+    def __enter__(self):
+        self.parent = self.stack.scope
+        self.stack.scope = self
+        return self
+
+    def __exit__(self, *args):
+        self.stack.scope = self.parent
+        return False
+
+    def is_global_scope(self):
+        return self.id == 0
+
+    def is_function_scope(self):
+        return isinstance(self, ScopeFn) and self.is_fn
+
+    def inside_function_scope(self):
+        cur = self
+        while cur.id != 0:
+            if cur.is_function_scope():
+                return True
+            cur = cur.parent
+        return False
+
+    def nearest_python_scope(self):
+        cur = self
+        while cur.id != 0 and isinstance(cur, ScopeLet):
+            cur = cur.parent
+        return cur
+
+    def access(self, node, index=None):
+        if not isinstance(node, NodeWithName):
+            node = NodeWithName(node, index)
+        self._access(node)
+
+    def assign(self, node, index=None):
+        if not isinstance(node, NodeWithName):
+            node = NodeWithName(node, index)
+        self._assign(node)
+
+    def _access(self, node):
+        pass
+
+    def _assign(self, node):
+        pass
+
+    def define(self, name):
+        pass
+
+    def unlocal(self, nonlocal_node):
+        if isinstance(nonlocal_node, ast.Nonlocal):
+            raise SyntaxError("cannot use nonlocal in this scope")
+
+
+# Used for lets and excepts
+class ScopeLet(Scope):
+    def __init__(self, stack, scope_id):
+        super().__init__(stack, scope_id)
+        self.bindings = {}
+
+    def _assign(self, node):
+        if node.name in self.bindings:
+            node.name = self.bindings[node.name]
+        else:
+            # we don't bind this name, toss it up to parent scope
+            self.parent.assign(node)
+
+    def _access(self, node):
+        if node.name in self.bindings:
+            node.name = self.bindings[node.name]
+        else:
+            # we don't bind this name, toss it up to parent scope
+            self.parent.access(node)
+
+    def add(self, target):
+        if isinstance(target, Symbol):
+            if "." in target:
+                raise ValueError("binding target may not contain a dot")
+            name = mangle(target)
+            new_name = f"_hy_scope_{self.id}_{name}"
+            self.bindings[name] = new_name
+            return Symbol(new_name).replace(target)
+        elif isinstance(target, List):
+            return List(map(self.add, target)).replace(target)
+        elif isinstance(target, Expression) and target:
+            if isinstance(target[0], Symbol) and str(target[0]) in (",", "unpack-iterable"):
+                return Expression([target[0], *map(self.add, target[1:])]).replace(target)
```

Comment:
this fails confusingly when attempting to use a star expression as the root symbol
```clojure
(let [#* a [1 2 3]] a)
```
`walk.let` fails like
```clojure
    (let [#* a [1 2 3]])
    ^------------------^
expanding macro let
  TypeError: bind targets must be symbol or iterable unpacking assignment
```